packages/react-router/tests/link.test-d.tsx

```diff
@@ -216,6 +230,116 @@ test('when navigating from a route with no params to a route with params, params
   params.parameter(0).toEqualTypeOf<{}>()
 })
 
+test('when navigating from a route to a route with the same params, the params can be an object with optional properties', () => {
+  const TestLink = Link<RouteTree, '/invoices/$invoiceId', './edit'>
+  const params = expectTypeOf(TestLink).parameter(0).toHaveProperty('params')
+
+  params
+    .exclude<Function | boolean>()
+    .branded.toEqualTypeOf<{ invoiceId?: string | undefined } | undefined>()
+})
+
+test('when navigating from a route to a route with the same params, params is a function from params to optional next params', () => {
+  const TestLink = Link<RouteTree, '/invoices/$invoiceId', './edit'>
+  const params = expectTypeOf(TestLink)
+    .parameter(0)
+    .toHaveProperty('params')
+    .extract<Function>()
+
+  params.returns.branded.toEqualTypeOf<{ invoiceId?: string | undefined }>()
+  params.parameter(0).toEqualTypeOf<{ invoiceId: string }>()
+})
+
+test('when navigating to a union of routes, params is optional', () => {
+  const TestLink = Link<
+    RouteTree,
+    string,
+    '/invoices/$invoiceId/' | '/posts/$postId/'
+  >
+  expectTypeOf(TestLink).parameter(0).not.toMatchTypeOf<{ params: unknown }>()
```

Comment:
is this the type check for "params is optional"? how? and is there a more explicit way?

Reply:
Docs cover this

https://github.com/mmkal/expect-type#error-messages

Reply:
I'm going to fix this comments tomorrow so don't merge. But you have me doubting myself so I will check

Reply:
I'm trying to think of a different way but it's difficult as the types can be quite complex for `params`